### PR TITLE
{Packaging} MSI: Use `WixBroadcastEnvironmentChange` and set `WIXUI_EXITDIALOGOPTIONALTEXT`

### PR DESCRIPTION
**Pull request by heaths**
**Description**
This is an internal ask from the CLI team to make some minimal changes regarding environment variables.

* Use built-in env var propagation
* Show message to restart terminals first time

**Testing Guide**
1. Install new MSI produced with changes. Verify message shows up on exit dialog.
2. Uninstall MSI.
3. Install older MSI (no message; wasn't added)
4. Install new MSI again. Verify no message because the old version already set the PATH.

Technically, someone could install the old version and turn right around and install the new one without restarting terminals, but that seems unlikely. It's also a point in time problem. Once the "older MSI" in the scenario above has this dialog change, it would've showed the message anyway. Thus, I don't recommend showing this message more often that not. It would be unnecessary for terminals to be restarted (potentially losing command history, session variables, etc.) if the PATH wasn't effectively changed during an upgrade.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
MSI finish page enhancement

File: build_scripts/windows/Product.wxs [hunk 1]
```diff
@@ -61,6 +51,12 @@
     <UIRef Id="WixUI_Minimal"/>
     <UIRef Id="WixUI_ErrorProgressText"/>
 
+    <!-- Show message to restart any terminals only if the PATH is changed -->
+    <CustomAction Id="Set_WIXUI_EXITDIALOGOPTIONALTEXT" Property="WIXUI_EXITDIALOGOPTIONALTEXT" Value="Please close and reopen any active terminal window to use Azure CLI." />
+    <InstallUISequence>
+      <Custom Action="Set_WIXUI_EXITDIALOGOPTIONALTEXT" After="CostFinalize">NOT Installed AND NOT WIX_UPGRADE_DETECTED</Custom>
+    </InstallUISequence>
```
Comment (jiasli) [inline build_scripts/windows/Product.wxs:58]: I saw this document: https://wixtoolset.org//documentation/manual/v3/wixui/wixui_customizations.html, and it uses

```xml
  <CustomAction Id="CA_Set_WIXUI_EXITDIALOGOPTIONALTEXT" Property="WIXUI_EXITDIALOGOPTIONALTEXT" Value="Thank you for installing [ProductName]."/>
  <InstallUISequence>
    <Custom Action="CA_Set_WIXUI_EXITDIALOGOPTIONALTEXT" After="FindRelatedProducts">NOT Installed</Custom>
  </InstallUISequence>
```
Just wondering what the difference between `CostFinalize` and `FindRelatedProducts` is.

**Comment by jiasli** [comment]
When I try to build MSI locally with

```
msbuild /t:rebuild /p:Configuration=Release %REPO_ROOT%\build_scripts\windows\azure-cli.wixproj
```

I got error message:

```
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 9/1/2022 4:15:50 PM.
Project "D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" on node 1 (rebuild target(s)).
CoreClean:
  Creating directory "out\obj\Release\".
D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj(56,5): error MSB4062: The "HeatDirectory" task could not be lo
aded from the assembly D:\cli\azure-cli\build_scripts\windows\artifacts\wix\wixtasks.dll. Could not load file or assemb
ly 'Microsoft.Build.Utilities, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its depende
ncies. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembl
y and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Fra
mework.ITask.
Done Building Project "D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" (rebuild target(s)) -- FAILED.


Build FAILED.

"D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" (rebuild target) (1) ->
(BeforeBuild target) ->
  D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj(56,5): error MSB4062: The "HeatDirectory" task could not be
loaded from the assembly D:\cli\azure-cli\build_scripts\windows\artifacts\wix\wixtasks.dll. Could not load file or asse
mbly 'Microsoft.Build.Utilities, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its depen
dencies. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assem
bly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.F
ramework.ITask.

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.21
```

According to https://github.com/wixtoolset/issues/issues/3872/#issue-123149061, [install .NET 3.5](https://www.microsoft.com/en-us/download/details.aspx?id=21) unblocks me.

**Comment by heaths** [comment]
> When I try to build MSI locally with
> 
> ```
> msbuild /t:rebuild /p:Configuration=Release %REPO_ROOT%\build_scripts\windows\azure-cli.wixproj
> ```
> 
> I got error message:
> 
> ```
> Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
> Copyright (C) Microsoft Corporation. All rights reserved.
> 
> Build started 9/1/2022 4:15:50 PM.
> Project "D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" on node 1 (rebuild target(s)).
> CoreClean:
>   Creating directory "out\obj\Release\".
> D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj(56,5): error MSB4062: The "HeatDirectory" task could not be lo
> aded from the assembly D:\cli\azure-cli\build_scripts\windows\artifacts\wix\wixtasks.dll. Could not load file or assemb
> ly 'Microsoft.Build.Utilities, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its depende
> ncies. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembl
> y and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Fra
> mework.ITask.
> Done Building Project "D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" (rebuild target(s)) -- FAILED.
> 
> 
> Build FAILED.
> 
> "D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj" (rebuild target) (1) ->
> (BeforeBuild target) ->
>   D:\cli\azure-cli\build_scripts\windows\azure-cli.wixproj(56,5): error MSB4062: The "HeatDirectory" task could not be
> loaded from the assembly D:\cli\azure-cli\build_scripts\windows\artifacts\wix\wixtasks.dll. Could not load file or asse
> mbly 'Microsoft.Build.Utilities, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its depen
> dencies. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assem
> bly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.F
> ramework.ITask.
> 
>     0 Warning(s)
>     1 Error(s)
> 
> Time Elapsed 00:00:00.21
> ```
> 
> According to [wixtoolset/issues#3872](https://github.com/wixtoolset/issues/issues/3872), [installing .NET 3.5](https://www.microsoft.com/en-us/download/details.aspx?id=21) unblocks me.

You have to build with the 32-bit msbuild. If you started a Developer Command Prompt from VS, it defaults to your machine bitness. That's not related to any of my changes, but something you might want to fix with your build scripts. Though, your build scripts seem to be right if you run then directly, but it isn't incremental.

File: build_scripts/windows/Product.wxs (hunk 1 above)
Comment (heaths) [reply build_scripts/windows/Product.wxs:58]: `CostFinalize` computers the final feature and component states. `FindRelatedProducts` should be scheduled before that, but feature and component states haven't been fully computer yet. `CostFinalize` is more typical.

It doesn't matter, though: both of these actions happen *way* before the end when the property actually gets used.

Speaking of `FindRelatedProducts`, you don't even need it and my other PR removes it. You have 1 feature. It's pointless and time-consuming to migrate feature states for 1 feature. If 1 feature is installed, the product is installed. If 0 features are installed, the product gets uninstalled. So you're not migrating anything; however, that action to actually process the old products' feature states and reapply them to the new product takes several seconds on my sandbox environment. That's time wasted.

**Review by jiasli — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by wangzelin007 — APPROVED** [review]

**Pull request merged**